lib/pos_consensus_controller.go

```diff
@@ -211,3 +211,109 @@ func (cc *ConsensusController) HandleHeader(pp *Peer, msg *MsgDeSoHeader) {
 func (cc *ConsensusController) HandleBlock(pp *Peer, msg *MsgDeSoBlock) {
 	// TODO
 }
+
+// processNewTipBlock tries to apply a new tip block to both the Blockchain and FastHotStuffEventLoop data
+// structures. It guarantees
+func (cc *ConsensusController) tryProcessBlockAsNewTip(block *MsgDeSoBlock) ([]*BlockHash, error) {
+	// Try to apply the block locally as the new tip of the blockchain
+	successfullyAppliedNewTip, _, missingBlockHashes, err := cc.blockchain.processBlockPoS(
+		block, // Pass in the block itself
+		cc.fastHotStuffEventLoop.GetCurrentView(), // Pass in the current view to ensure we don't process a stale block
+		true, // Make sure we verify signatures in the block
+	)
+	if err != nil {
+		return nil, errors.Errorf("HandleFastHostStuffBlockProposal: Error processing block locally: %v", err)
+	}
+
+	// If the incoming block is an orphan, then there's nothing we can do. We return the missing ancestor
+	// block hashes to the caller. The caller can then fetch the missing blocks from the network and retry
+	// if needed.
+	if len(missingBlockHashes) > 0 {
+		return missingBlockHashes, nil
+	}
+
+	// If the block was processed successfully but was not applied as the new tip, then there's nothing left to do.
+	if !successfullyAppliedNewTip {
+		return nil, nil
+	}
+
+	// Fetch the new tip from the blockchain. Note: the new tip may or may not be the input block itself.
+	// It's possible that there was a descendant of the tip block that was previously stored as an orphan
+	// in the Blockchain, and was applied as the new tip.
+	tipBlockHeader := cc.blockchain.BlockTip().Header
+	tipBlockHash, err := tipBlockHeader.Hash()
+	if err != nil {
+		return nil, errors.Errorf("HandleFastHostStuffBlockProposal: Error hashing new tip block: %v", err)
+	}
+
+	// Build a UtxoView at the new tip
+	utxoViewAtTipBlock, err := cc.blockchain.getUtxoViewAtBlockHash(*tipBlockHash)
+	if err != nil {
+		return nil, errors.Errorf("HandleFastHostStuffBlockProposal: Error fetching UtxoView for tip block: %v", err)
+	}
+
+	// Fetch the validator set at the new tip
+	validatorSetAtTipBlock, err := utxoViewAtTipBlock.GetFullSnapshotValidatorSet()
```

Comment:
hmm do we want the validator set from before connecting the block? In `pos_blockchain.go`, we connect to the parent block and then pull the validator set before connecting the incoming block. I'm a little concerned about the edge case where we have something like the below :

```
| ------- epoch n ------- | ------ epoch n + 1 ------ |
                       ... x-1 -> x -> x + 1
```
Where x+1 is the first block in the next epoch. When we get the utxo view at X, GetCurrentEpochEntry will return epoch n + 1.

Reply:
No, we want the validator set after connecting the block. The way to think about the validator set is "this is the validator set we use to validate the next block that extends from this safe block".

Reply:
ahhhh okay. Maybe we can clarify that in a comment somewhere? or fix the naming a bit? Not sure what makes it most clear.

Reply:
I added comments for this where we fetch the validator sets for the tip block and the safe blocks